Box provider implementation

cartfisk opened the pull request:
Hey there! I'm an engineer at [MediaOcean](https://mediaocean.com) and have had my eyes on Uppy for a little while.

Coincidentally, I got around to a proper investigation this week as part of our annual hackathon and confirmed that Uppy makes an excellent fit into a couple file-heavy MediaOcean products (minus the fact that it is missing Box support), so I figured I'd take a stab at contributing.

~~This PR is arriving **unfinished**. I'm hoping to get a quick WIP once-over by someone on the team, primarily because I cannot, for the life of me, figure out how to get thumbnails working (along with a number of additional issues I'm sure to discover after thumbnails are resolved).~~

~~Box doesn't provide public thumbnails, so I followed the Dropbox provider as an example. From what I can tell, the `thumbnail` URL injected by `adaptData()` into `list` results isn't resolving correctly to my provider's `thumbnail()` method.~~

Hope this is useful! Hats off to the awesome React integration, easy contribution workflow, helpful docs, and readable, clean code. 10/10.

Cheers!

---
https://github.com/transloadit/uppy/issues/2183

arturi commented:
Hi! Wow, thank you! 🤩 This is great news, Box was on our todo-list, this helps a lot. Ife, who is working on Companion in our team, is currently on vacation (till the end of the month). Once he’s back, he’ll be able to take a close look and help you out.

cartfisk commented:
Thanks for the follow-up. Hopefully I can make some more progress on my own by the time Ife is back.

cartfisk commented:
Updating as I go– the issue actually seems to be related to token passthrough to the `thumbnails` route. The `verifyToken` middleware can't find `authToken` on the `companion` instance and never passes the request through to `thumbnails()`.

cartfisk commented:
Seems like everything is working locally for me now. Added docs and cleaned things up a bit. The tests still need some work (missing some context on mock credentials and E2E architecture), and someone will need to configure a Box developer app for the examples site.

ifedapoolarewaju reviewed (commented):
Thank you so much for the work! This is really great work, I only had a few comments

cartfisk commented on packages/@uppy/companion/src/server/provider/box/index.js:
```diff
@@ -0,0 +1,176 @@
+      .auth(token)
+      .request()
+      .on('response', (resp) => {
+        if (![200, 202].includes(resp.statusCode)) {
```
Not sure if it's an API quirk– thumbnails came back as documented for the 200 case but with the 202 code (at least initially, only allowing 200 here was swallowing the responses). I didn't need to access headers for a placeholder, the thumbnail came back as expected for a 200.

This is maybe the first time I've come across a 202 in practice, so for full disclosure, this was a moment of ¯\_(ツ)_/¯ it works.

cartfisk commented:
@ifedapoolarewaju Thanks for the thorough review– glad it meets the mark :) I'll touch up the straightforward things in the next couple days. Let me know if you have any ideas around the 202 issue, the most I can say is that it works as-is, but I don't know exactly why.

Cheers!

cartfisk commented:
@ifedapoolarewaju Besides the 202 question, the other marked issues have been resolved.

cartfisk commented:
@ifedapoolarewaju @arturi The last remaining opens are 
1) setting up Box application credentials for the examples site
2) E2E testing– I copied most of this from Dropbox, last time I looked at the PR check logs it looked like the tests were failing due to missing secrets but there may be some other issues, that area is cloudy for me still.

PR checks should pass as soon as these two things are resolved.

ifedapoolarewaju commented on packages/@uppy/companion/src/server/Uploader.js:
```diff
@@ -471,6 +478,9 @@ class Uploader {
     const headers = headerSanitize(this.options.headers)
     const reqOptions = { url: this.options.endpoint, headers, encoding: null }
     const httpRequest = request[httpMethod]
+    if (this.options.insecureUpload) {
```
what is this option for? 🤔

cartfisk replied:
Apologies, I should have diffed this closer, that was a little hack to make this work on our application's dev-server that operates with an expired cert. Not supposed to be included. I'll do another quick look over.

cartfisk replied:
@ifedapoolarewaju Pulled out the offending commit.

ifedapoolarewaju commented on website/src/examples/markdown-snippets/app.es6:
```diff
@@ -136,7 +136,8 @@ class MarkdownTextarea {
         'instagram',
         'google-drive',
-        'dropbox'
+        'dropbox',
+        'box'
```
excluding from examples in the meantime
```suggestion
        'dropbox'
```

ifedapoolarewaju commented on website/src/examples/dashboard/index.ejs:
```diff
@@ -26,6 +26,7 @@ const Uppy = require('@uppy/core')
 const Dashboard = require('@uppy/dashboard')
 const GoogleDrive = require('@uppy/google-drive')
 const Dropbox = require('@uppy/dropbox')
+const Box = require('@uppy/box')
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on website/src/examples/dashboard/app.html:
```diff
@@ -14,6 +14,7 @@
     <li><label for="opts-ScreenCapture"><input type="checkbox" id="opts-ScreenCapture" checked/> Screen Capture</label></li>
     <li><label for="opts-GoogleDrive"><input type="checkbox" id="opts-GoogleDrive" checked/> Google Drive</label></li>
     <li><label for="opts-Dropbox"><input type="checkbox" id="opts-Dropbox" checked/> Dropbox</label></li>
+    <li><label for="opts-Box"><input type="checkbox" id="opts-Box" checked/> Box</label></li>
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on website/src/examples/dashboard/app.html:
```diff
@@ -43,6 +44,7 @@
     ScreenCapture: document.querySelector('#opts-ScreenCapture'),
     GoogleDrive: document.querySelector('#opts-GoogleDrive'),
     Dropbox: document.querySelector('#opts-Dropbox'),
+    Box: document.querySelector('#opts-Box'),
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on website/src/examples/dashboard/app.html:
```diff
@@ -61,6 +63,7 @@
     GoogleDrive: true,
     Instagram: true,
     Dropbox: true,
+    Box: true,
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on website/src/examples/dashboard/app.es6:
```diff
@@ -103,6 +104,14 @@ function uppySetOptions () {
     window.uppy.removePlugin(dropboxInstance)
   }
 
+  const boxInstance = window.uppy.getPlugin('Box')
+  if (opts.Box && !boxInstance) {
+    window.uppy.use(Box, { target: Dashboard, companionUrl: COMPANION })
+  }
+  if (!opts.Box && boxInstance) {
+    window.uppy.removePlugin(boxInstance)
+  }
+
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on website/src/examples/dashboard/app.es6:
```diff
@@ -4,6 +4,7 @@ const Uppy = require('@uppy/core')
 const Dashboard = require('@uppy/dashboard')
 const GoogleDrive = require('@uppy/google-drive')
 const Dropbox = require('@uppy/dropbox')
+const Box = require('@uppy/box')
```
excluding from examples in the meantime
```suggestion
```

ifedapoolarewaju commented on packages/uppy/index.mjs:
```diff
@@ -20,6 +20,7 @@ export { default as StatusBar } from '@uppy/status-bar'
 
 // Acquirers
 export { default as Dropbox } from '@uppy/dropbox'
+export { default as Box } from '@uppy/box'
```
excluding box from global module for now as it will initially be released as beta
```suggestion
```

ifedapoolarewaju commented on packages/uppy/index.js:
```diff
@@ -22,6 +22,7 @@ exports.StatusBar = require('@uppy/status-bar')
 
 // Acquirers
 exports.Dropbox = require('@uppy/dropbox')
+exports.Box = require('@uppy/box')
```
excluding box from global module for now as it will initially be released as beta

```suggestion
```

ifedapoolarewaju commented on packages/@uppy/robodog/src/addProviders.js:
```diff
@@ -3,6 +3,7 @@ const has = require('@uppy/utils/lib/hasProperty')
 
 const remoteProviders = {
   dropbox: require('@uppy/dropbox'),
+  box: require('@uppy/box'),
```
excluding box from global module for now as it will initially be released as beta
```suggestion
```

ifedapoolarewaju commented on packages/@uppy/robodog/package.json:
```diff
@@ -31,6 +31,7 @@
     "@uppy/core": "file:../core",
     "@uppy/dashboard": "file:../dashboard",
     "@uppy/dropbox": "file:../dropbox",
+    "@uppy/box": "file:../box",
```
excluding box from global module for now as it will initially be released as beta
```suggestion
```
